### PR TITLE
[CALCITE-6313] Add POWER function for PostgreSQL

### PR DESCRIPTION
**Pull request by normanj-bitquill**
* The existing power function is moved to all non PostgreSQL libraries
* The new power function is only for PostgreSQL
* The new function returns a decimal if any argument is a decimal

### CODE REVIEW

**Comment by normanj-bitquill** [comment]
In order to separate the POWER function for PostgreSQL and non-PostgreSQL dialects, the POWER function was moved to the `SqlLibraryOperators` class. This downgrades the POWER function from a standard function to a function from a library. This is problematic for things like SQRT which is a standard function and gets rewritten to a POWER expression.

Is there a more clean way to separate the POWER function so that the PostgreSQL implementation is able to return Decimal values?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2241]: this comment seems wrong

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:413]: this comment seems wrong too

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:794]: why have both DOUBLE_OR_DECIMAL and DECIMAL_OR_DOUBLE?

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:562]: why delete these tests? What happens to them? They look like good tests.

File: core/src/test/resources/sql/misc.iq
Comment (mihaibudiu) [inline core/src/test/resources/sql/misc.iq:2304]: why delete this test?

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2009]: these quotes look suspicious

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Updated the comment.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java]: Moved the comment to `DECIMAL_OR_DOUBLE`. Removed `DOUBLE_OR_DECIMAL`.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:780]: Removed `DOUBLE_OR_DECIMAL`.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
Comment (normanj-bitquill) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:562]: The `POWER` function is no longer available in `SqlStdOperatorTable`. The implementation to use depends on which library is active since PostgreSQL can use a different return type.

This test fails as it was since `POWER` is not in `StdOperatorTable`. Is there a way to specify the library to use for the test?

Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:562]: Yes, see the operatorTableFor() function.

Comment (normanj-bitquill) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:562]: Updated.

File: core/src/test/resources/sql/misc.iq
Comment (normanj-bitquill) [reply core/src/test/resources/sql/misc.iq:2304]: The `POWER` function is no longer available in `SqlStdOperatorTable`. Is there any way of specifying a library to use here?

Comment (mihaibudiu) [reply core/src/test/resources/sql/misc.iq:2304]: I don't think there is, but the question is whether there should be a standard power function.

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
Comment (normanj-bitquill) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2009]: This will end up going through here:
https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql/SqlUtil.java#L393

and only look in `SqlStdOperatorTable` for the function.

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2009]: Then it's related to my previous question: should there be a variant of this function in the standard operator table?
If not, I guess this is fine.

Comment (normanj-bitquill) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2009]: Is it possible to have a power function in the standard operator table and still have an override for PostgreSQL that is able to return a decimal in some cases?

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:2009]: I don't know the answer to this question. But I know that Calcite can load multiple libraries, and I believe a function will be resolved by trying the libraries in order. I also am assuming that there is nothing special about the standard operator table. So if you load the postgres operator table first, perhaps the right POWER will be used?

A quick experiment should be able answer this question.

If this works, we should probably document that the right way to implement a dialect is to load the dialect specific operator table first.

Comment (normanj-bitquill) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java]: I found where the `ChainedSqlOperatorTable` is created and was able to give the library operator table precedence over the standard operator table.

With this change, no changes are needed to existing tests.

**Comment by mihaibudiu** [comment]
In general, until the changes are approved I would recommend adding new commits instead of force-pushing. It makes it easier for reviewers to see what's new. Then you can squash when the PR is approved.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperatorTableFactory.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperatorTableFactory.java:120]: This may cause problems for other people. Ideally you want to do this just for your tests.

Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperatorTableFactory.java:120]: Understood. I'll take a look.

Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperatorTableFactory.java]: This has been fixed. The only changes needed now are in the unit test itself. I setup the operator table there such that PostgreSQL is first.

**Review by mihaibudiu — APPROVED** [review]
The second commit should have a descriptive name, e.g., "only update loading library order for tests." But that is moot now, since I think this is ready for merging, so the next step is to squash the commits. We can't merge this until 1.37 is released, though.

**Pull request merged**
